test/extensions/filters/network/kafka/broker/filter_integration_test.cc

```diff
@@ -0,0 +1,155 @@
+#include "common/common/utility.h"
+#include "common/stats/isolated_store_impl.h"
+
+#include "extensions/filters/network/kafka/broker/filter.h"
+#include "extensions/filters/network/kafka/external/requests.h"
+#include "extensions/filters/network/kafka/external/responses.h"
+
+#include "test/extensions/filters/network/kafka/buffer_based_test.h"
+#include "test/extensions/filters/network/kafka/message_utilities.h"
+#include "test/test_common/test_time.h"
+
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace Kafka {
+namespace Broker {
+
+using RequestB = MessageBasedTest<RequestEncoder>;
+using ResponseB = MessageBasedTest<ResponseEncoder>;
+
+// Message size for all kind of broken messages (we are not going to process all the bytes).
+constexpr static int32_t BROKEN_MESSAGE_SIZE = std::numeric_limits<int32_t>::max();
+
+class KafkaBrokerFilterIntegrationTest : public testing::Test,
```

Comment:
This is not really an integration test in the normal sense of the way we use this term (real networking, fake clients, fake servers, etc.). Can you rename to be more clear about what this is? Optimally you would have a real end-to-end integration test for this filter due to the complexity. Is that possible?

Reply:
I think so.
If I understand it correctly, I have two ways to test it - A) proper integration test with fake upstreams mocking kafka server behaviour (similar to `redis_proxy_integration_test.cc`) OR B) full setup of environment with kafka & some simple kafka client. 

WDYT about B) implemented in such a way:
```
- download kafka from apache site and start it
- run the envoy that's being tested right now
- run the tests (that might mean invoking external applications (java kafka clients))
[basically what I do when I develop this filter, just in automated fashion]
```
do you know if it would be possible to hook this kind of process into `bazel test`? or would we need extra wiring?

Reply:
cc @zuercher who did integration tests for Thrift and should have some good ideas.

Reply:
I did take a look but (if I'm reading all of this correctly) all integration tests depend on `FakeUpstream` that would need to be fed with real Kafka protocol payloads
(and here I don't see too much benefit, as effectively we'd be re-implementing Kafka servers' behaviour).

But!
I think I can simply create another (not necessarily bash) script like https://github.com/envoyproxy/envoy/blob/master/test/exe/envoy_static_test.sh in kafka part, and here I might be capable of rolling out the full test, driven by java kafka client.

(The main motivator to use real kafka is that (for example) simple sending a message from KafkaProducer includes 1 api_versions_request + 1 metadata_request + 1 produce_request - for more complex client APIs it might be even worse.)

WDYT?

Reply:
I think this would be optimal, but we can defer this until later if you want.

Reply:
Renamed to `ProtocolTests` as we are testing all types of messages supported by the protocol.
Proper integration test was added in `kafka/broker/integration_test` (basically we start kafka+zk+envoy and run python-kafka client with kafka operations pointing to envoy).